WikiWatershed/model-my-watershed
# Ensure Celery tasks get routed to matching worker

## Overview

Configure Celery to only send tasks to workers that match the specified stack color. This prevents tasks from getting assigned to an old worker in a blue/green deployment.

Configuration is largely based on http://docs.celeryproject.org/en/latest/userguide/routing.html#manual-routing and https://github.com/rduplain/celery-simple-example.

Connects to #2374

## Testing Instructions

- **Before pulling down this branch**, make sure your worker VM is up.
- Next, **still don’t pull down this branch**, but apply the attached patch to bring up another worker. [pr.patch.zip](https://github.com/WikiWatershed/model-my-watershed/files/1422474/pr.patch.zip)
- When both workers are running, in two separate terminal windows (it’s easier to see if it’s one window split into two panes), run both `./scripts/debugcelery.sh` and `./scripts/debugworkerm.sh`.
- Start the app using `./scripts/debugserver.sh`.
- Select an AoI, and switch back to the terminal. You should see tasks being routed to both workers.
- Select MapShed, and switch back to the terminal. Again, you should see output from both workers because the tasks for MapShed are being split amongst both of them.
- Now, pull down this branch. 
- Restart the debug scripts for both workers. You should note that, in the start up text, you see output like:

```
> celery           exchange=direct key=task.{Color}
````

Previously, it looked something like:

```
> celery           exchange=celery(direct) key=celery
````
- The original worker should output `key=task.Black`, and the new worker should output `key=task.Magenta`.
- Return to the app, and kick off a mashed job. Only the Magenta worker should get tasks.
- Edit [Line 73](https://github.com/WikiWatershed/model-my-watershed/blob/develop/src/mmw/mmw/settings/base.py#L73) of the settings file and change the value to `Black`. Restart the app server.
- Kick off another modeling job. Tasks should now only be sent to the Black worker.
